pym/bob/scm/git.py

```diff
@@ -253,6 +253,15 @@ async def __checkoutTag(self, invoker, fetchCmd, switch):
                         self.fail("Plain git-fetch in", self.__dir,
                             "did not download the requested commit and the explicit fetch failed!",
                             returncode=ok)
+
+            # do not perform inline switch if we're in a unclean detaced
+            # head state
+            status = await invoker.checkOutputCommand(["git",  "status", "--porcelain"],
+                    cwd=self.__dir)
+            if status:
+                await invoker.checkCommand(["git", "symbolic-ref",
+                    "-q", "HEAD"], cwd=self.__dir)
```

Comment:
What about a clean detached head state? I thought the original issue was that somebody made commits without being on a branch..?

I would not make any distinction about clean/unclean because git handles this if modified files would be overwirtten. Unrelated files can still be left as-is. There is already a dedicated switch to ensure clean checkouts.

I think this is unrelated to checkouts of tags/commits. The user could have checked out a tag/commit himself. How about the following: If we ever switch away from a detached-head state then move to attic. The only exception is that the old checkout (`oldSpec`) was on the commit or tag that still matches the current HEAD. This would retain switching from tag/commits to other states unless the user checked out some other commit or made commits on detached heads. If the user created a branch we would also do an inline switch which is IMHO ok. What do you think?

Reply:
> What about a clean detached head state? I thought the original issue was that somebody made commits without being on a branch..?
> 
> I would not make any distinction about clean/unclean because git handles this if modified files would be overwirtten. Unrelated files can still be left as-is. There is already a dedicated switch to ensure clean checkouts.
> 
A clean detached state is the state we have with a `commit` checked out. I don't want to move to `attic` every time the commit ID is updated without a reason.

> I think this is unrelated to checkouts of tags/commits. The user could have checked out a tag/commit himself. How about the following: If we ever switch away from a detached-head state then move to attic. The only exception is that the old checkout (`oldSpec`) was on the commit or tag that still matches the current HEAD. This would retain switching from tag/commits to other states unless the user checked out some other commit or made commits on detached heads. If the user created a branch we would also do an inline switch which is IMHO ok. What do you think?

:+1: sounds good.